File: api/routes/authentication.js
const express = require('express');
const router = express.Router();

const {
  login: loginValidation,
  register: registerValidation,
} = require('../validation/authentication');

const { register, login } = require('../services/authentication');

router.post('/login', async (req, res) => {
  const error = loginValidation(req.body);
  if (error) {
    const message = error.replace(/"/g, '');
    return res.status(400).json({ message, status: 400 });
  }
  const { status, message, result } = await login(req.body);
  res.status(status).json({ message, result, status });
});

router.post('/sign_up', async (req, res) => {
  const error = registerValidation(req.body);
  if (error) {
    const message = error.replace(/"/g, '');
    return res.status(400).json({ message, status: 400 });
  }
  const { status, message } = await register(req.body);

  res.status(status).json({ message,  status });
});

module.exports = router;

File: mobile_app/lib/services/authentication.dart
import 'dart:convert';
import 'dart:io';

import 'package:mobile_app/config/constants.dart';
import 'package:mobile_app/models/authentication/login.dart';
import 'package:mobile_app/models/authentication/register.dart';
import 'package:mobile_app/services/bearer_token.dart';
import 'package:http/http.dart' as http;
import 'package:mobile_app/services/response_parser.dart';

class AuthenticationService {
  final BearerTokenService bearerTokenService;

  AuthenticationService({required this.bearerTokenService});
  Future<LoginResponse> login(Map<String, String> body) async {
    try {
      final baseUrl = '${Constants.baseUrl}/login';

      http.Response response = await http.post(
        Uri.parse(baseUrl),
        headers: getHeaders(),
        body: jsonEncode(body),
      );
      return ResponseParserUtil.parseLogin(response.body);
    } on SocketException {
      return LoginResponse(status: 500, message: 'No Internet connection');
    }
  }

  Future<RegistrationResponse> register(Map<String, String> body) async {
    try {
      final baseUrl = '${Constants.baseUrl}/sign_up';
      http.Response response = await http.post(
        Uri.parse(baseUrl),
        headers: getHeaders(),
        body: jsonEncode(body),
      );
      return ResponseParserUtil.parseLogin(response.body);
    } on SocketException {
      return RegistrationResponse(
          status: 500, message: 'No Internet connection');
    }
  }

  Map<String, String> getHeaders() {
    Map<String, String> headers = {
      'Content-Type': 'application/json; charset=UTF-8',
      'Accept': 'application/json',
    };
    String token = bearerTokenService.get();

    if (token.isNotEmpty) {
      headers['Authorization'] = 'Bearer $token';
    }
    return headers;
  }
}
